pkg/controller/dynamicupgradeableconditionscontroller.go

```diff
@@ -0,0 +1,300 @@
+	"github.com/openshift/library-go/pkg/controller/factory"
+	"github.com/openshift/library-go/pkg/operator/events"
+	"github.com/openshift/library-go/pkg/operator/v1helpers"
+	helmclient "github.com/operator-framework/helm-operator-plugins/pkg/client"
+	storage "github.com/operator-framework/helm-operator-plugins/pkg/storage"
+	ocv1alpha1 "github.com/operator-framework/operator-controller/api/v1alpha1"
+	"github.com/operator-framework/operator-registry/alpha/property"
+	helm "helm.sh/helm/v3/pkg/storage"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/dynamic/dynamicinformer"
+	"k8s.io/client-go/informers"
+	"k8s.io/client-go/kubernetes"
+	v1 "k8s.io/client-go/kubernetes/typed/core/v1"
+	"k8s.io/client-go/tools/cache"
+)
+
+const (
+	incompatibleOperatorsInstalled = "IncompatibleOperatorsInstalled"
+	MaxOpenShiftVersionProperty    = "olm.maxOpenShiftVersion"
+	OwnerKindKey                   = "olm.operatorframework.io/owner-kind"
+	OwnerNameKey                   = "olm.operatorframework.io/owner-name"
+	PackageNameKey                 = "olm.operatorframework.io/package-name"
+	BundleNameKey                  = "olm.operatorframework.io/bundle-name"
+	BundleVersionKey               = "olm.operatorframework.io/bundle-version"
+)
+
+type dynamicUpgradeableConditionController struct {
+	kubeclient     kubernetes.Interface
+	informer       informers.GenericInformer
+	client         dynamic.Interface
+	name           string
+	operatorClient *clients.OperatorClient
+	prefixes       []string
+}
+
+type ChartMetadata struct {
+	Metadata struct {
+		Annotations map[string]string `yaml:"annotations"`
+	} `yaml:"metadata"`
+}
+
+type SecretData struct {
+	Chart struct {
+		ChartMetadata
+	} `yaml:"chart"`
+}
+
+func NewDynamicUpgradeableConditionController(kubeclient kubernetes.Interface, client dynamic.Interface, name string, operatorClient *clients.OperatorClient, eventRecorder events.Recorder, prefixes []string) factory.Controller {
+	infFact := dynamicinformer.NewDynamicSharedInformerFactory(client, 0)
+
+	clusterExtensionGVR := schema.GroupVersionResource{
+		Group:    ocv1alpha1.ClusterExtensionGVK.Group,
+		Version:  ocv1alpha1.ClusterExtensionGVK.Version,
+		Resource: ocv1alpha1.ClusterExtensionGVK.Kind,
+	}
+
+	inf := infFact.ForResource(clusterExtensionGVR)
+
+	c := &dynamicUpgradeableConditionController{
+		informer:       inf,
+		client:         client,
+		kubeclient:     kubeclient,
+		name:           name,
+		operatorClient: operatorClient,
+		prefixes:       prefixes,
+	}
+
+	inf.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    c.handleClusterExtension,
+		UpdateFunc: func(oldObj, newObj interface{}) { c.handleClusterExtension(newObj) },
+		DeleteFunc: c.handleClusterExtension,
+	})
+
+	return factory.New().WithInformers(inf.Informer()).ToController(name, eventRecorder)
```

Comment:
Looking into the `github.com/openshift/library-go/pkg/controllers/factory` package, it looks like using `WithInformers()` is for setting informers that you want to trigger the `Sync` function for this controller [1]. If you want to use custom event handlers, you probably want to use `WithBareInformers()` [2].

[1]: https://github.com/openshift/library-go/blob/7c5c5d305ac1bb911c5a4a0be82fa7d5aa7c5e35/pkg/controller/factory/factory.go#L95-L101
[2]: https://github.com/openshift/library-go/blob/7c5c5d305ac1bb911c5a4a0be82fa7d5aa7c5e35/pkg/controller/factory/factory.go#L115-L123